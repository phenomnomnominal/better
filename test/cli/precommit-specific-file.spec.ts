import simpleGit from 'simple-git';

import { createFixture } from '../fixture';

const ARGV = ['node', './bin/betterer'];

describe('betterer precommit', () => {
  it('should test just the specified files', async () => {
    const { paths, logs, cleanup, resolve, readFile, writeFile } = await createFixture('precommit-specific-file', {
      '.betterer.js': `
const { eslint } = require('@betterer/eslint');

module.exports = {
  test: () => eslint({ 'no-debugger': 'error' }).include('./src/**/*.ts')
};
      `,
      '.eslintrc.js': `
const path = require('path');

module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 2018,
    project: path.resolve(__dirname, './tsconfig.json'),
    sourceType: 'module'
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking'
  ],
  rules: {
    'no-debugger': 1
  }
};
      `,
      'tsconfig.json': `
{
  "extends": "../../tsconfig.json",
  "include": ["./src/**/*", "./.betterer.js", "./.eslintrc.js"]
}
      `,
      './src/existing-file-1.ts': `
debugger;
      `,
      './src/existing-file-2.ts': `
debugger;
      `
    });

    const fixturePath = paths.cwd;
    const resultsPath = paths.results;

    const newFilePath = resolve('./src/new-file.ts');

    await writeFile(newFilePath, 'debugger;');

    const { cli__ } = await import('@betterer/cli');

    await cli__(fixturePath, [...ARGV, 'start'], false);

    await cli__(fixturePath, [...ARGV, 'precommit', newFilePath], false);

    expect(logs).toMatchSnapshot();

    const result = await readFile(resultsPath);

    expect(result).toMatchSnapshot();

    const git = simpleGit();
    await git.init();
    const status = await git.status([paths.results]);
    const [stagedResultsPath] = status.staged;
    expect(stagedResultsPath).toMatchSnapshot();

    await git.reset([resultsPath]);

    await cleanup();
  });
});
