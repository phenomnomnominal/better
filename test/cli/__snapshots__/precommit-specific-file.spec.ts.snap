// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer precommit should test just the specified files 1`] = `
Array [
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ test: \\"test\\" got checked for the first time! (3 issues) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! (3 issues) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! (3 issues) 🎉

1 test got checked. 🤔
1 test got checked for the first time! 🎉
",
  "Checking 1 file... 🤔

・ <project>/fixtures/precommit-specific-file/src/new-file.ts

🌟 Betterer:
",
  "Checking 1 file... 🤔

・ <project>/fixtures/precommit-specific-file/src/new-file.ts

🌟 Betterer: 1 test running...
🤔 test: running \\"test\\"!
",
  "Checking 1 file... 🤔

・ <project>/fixtures/precommit-specific-file/src/new-file.ts

🌟 Betterer: 1 test running...
✅ test: \\"test\\" stayed the same. (3 issues) 😐
",
  "Checking 1 file... 🤔

・ <project>/fixtures/precommit-specific-file/src/new-file.ts

🎉 Betterer: 1 test done!
✅ test: \\"test\\" stayed the same. (3 issues) 😐
",
  "Checked 1 file! 🔍

・ <project>/fixtures/precommit-specific-file/src/new-file.ts

🎉 Betterer: 1 test done!
✅ test: \\"test\\" stayed the same. (3 issues) 😐

1 test got checked. 🤔
1 test stayed the same. 😐
",
]
`;

exports[`betterer precommit should test just the specified files 2`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`test\`] = {
  value: \`{
    \\"src/existing-file-1.ts:3201740415\\": [
      [0, 0, 9, \\"Unexpected \\\\'debugger\\\\' statement.\\", \\"3201740415\\"]
    ],
    \\"src/existing-file-2.ts:3201740415\\": [
      [0, 0, 9, \\"Unexpected \\\\'debugger\\\\' statement.\\", \\"3201740415\\"]
    ],
    \\"src/new-file.ts:3201740415\\": [
      [0, 0, 9, \\"Unexpected \\\\'debugger\\\\' statement.\\", \\"3201740415\\"]
    ]
  }\`
};
"
`;

exports[`betterer precommit should test just the specified files 3`] = `undefined`;
